src/snowflake/snowpark/mock/_stage_registry.py

```diff
@@ -0,0 +1,232 @@
+        for i, c in enumerate(normalized):
+            if c == '"':
+                is_quoted = (
+                    not is_quoted
+                )  # this handles escaping consecutive double quotes
+            elif c == "/" and not is_quoted:
+                # all chars prior to ith char is part of the stage name
+                stage_name_end_idx = i - 1
+                prefix_start_idx = i + 1
+                break
+
+        if not stage_name_end_idx:
+            raise SnowparkSQLException(f"Invalid stage_location {stage_location}.")
+    else:
+        stage_name_end_idx = normalized.find("/")
+        prefix_start_idx = stage_name_end_idx + 1
+    stage_name = normalized[stage_name_start_idx:stage_name_end_idx]
+    dir_path = normalized[prefix_start_idx:-1]  # remove the first and last '/'
+    return stage_name, dir_path
+
+
+class StageEntity:
+    def __init__(self, root_dir_path: str, stage_name: str) -> None:
+        self._stage_name = stage_name
+        # stage name might contain special chars which can not be used as dir name
+        # so we generate uuid as name
+        self._dir_name = str(uuid.uuid4())
+        self._working_directory = os.path.join(root_dir_path, self._dir_name)
+
+        if os.path.exists(self._working_directory):
+            shutil.rmtree(self._working_directory)
+
+        os.mkdir(self._working_directory)
+        self._files = set()
+
+    def put_file(
+        self, local_file_name: str, stage_prefix: str, overwrite: bool = False
+    ) -> TableEmulator:
+        local_file_name = local_file_name[
+            len("`file://") : -1
+        ]  # skip normalized prefix `file:// and suffix `
+        # glob supports wildcard '?' and '*' searching
+        list_of_files = glob.glob(local_file_name)
+        result_df = TableEmulator(
+            columns=PUT_RESULT_KEYS,
+            sf_types={
+                "source": ColumnType(StringType(), True),
+                "target": ColumnType(StringType(), True),
+                "source_size": ColumnType(DecimalType(10, 0), True),
+                "target_size": ColumnType(DecimalType(10, 0), True),
+                "source_compression": ColumnType(StringType(), True),
+                "target_compression": ColumnType(StringType(), True),
+                "status": ColumnType(StringType(), True),
+                "message": ColumnType(StringType(), True),
+            },
+            dtype=object,
+        )
+
+        if not list_of_files:
+            raise SnowparkSQLException(f"File doesn't exist: {local_file_name}")
+
+        for local_file_name in list_of_files:
+
+            file_name = os.path.basename(local_file_name)
+            stage_target_dir_path = os.path.join(self._working_directory, stage_prefix)
+
+            if not os.path.exists(stage_target_dir_path):
+                os.makedirs(stage_target_dir_path)
+
+            if (
+                os.path.isfile(os.path.join(stage_target_dir_path, file_name))
```

Comment:
NIT: could this path be an existing directory?

Reply:
nice catch, I tested and it can be a directory.
this is actually a problem to at least MacOS as it doesn't allow file and dir share the same name under a directory.

I have updated the implementation to append a suffix to each file to differentiate from a dir to bypass the issue.

Reply:
Does this mean a `test.py` PUT to a stage will not have `.py` as the suffix?

Reply:
updated to remove suffix